libnetwork/endpoint.go

```diff
@@ -631,6 +621,27 @@ func (ep *Endpoint) UpdateDNSNames(dnsNames []string) error {
 	return nil
 }
 
+func (ep *Endpoint) updateHostsFile(sb *Sandbox) error {
+	ep.mu.Lock()
+	n := ep.network
+	ep.mu.Unlock()
```

Comment:
It looks like these locks are new; I know there's a lot of "dubious" code either handling too many locks, or too little. While I do think that the signature is _nicer_ (encapsulating the locking), we should also consider that this function is not exported; for non-exported functions it's fine to specify additional requirements on callers (such as "caller must keep a lock").

- did we need the additional synchronisation here?
- I see the calling code (`sbJoin`) memoizes the network (has a local `n` variable); would passing the network as argument avoid needing the extra synchronisation?


I noitice that both `getFirstInterfaceIPv4Address` and `getFirstInterfaceIPv6Address` have locking, which in this case looks like it's actually adding complication (we must make sure to unlock before calling those), but ` sb.updateHostsFile` does not have any synchronisation.


Also, at a quick glance, `getFirstInterfaceIPv4Address` and `getFirstInterfaceIPv6Address` are not used anywhere  else (they're ONLY used here?)

If that's indeed the case, and looking at those functions, perhaps we should just inline them. We can also consider using an early return for the network-check;

```go
func (ep *Endpoint) updateHostsFile(sb *Sandbox) error {
	ep.mu.Lock()
	defer ep.mu.Unlock()

	// Do not update hosts file with internal network's endpoint IP
	if n := ep.network; n == nil || n.ingress || n.Name() == libnGWNetwork {
		return nil
	}

	var addresses []string
	if ep.iface.addr != nil {
		addresses = append(addresses, ep.iface.addr.String())
	}
	if ep.iface.addrv6 != nil {
		addresses = append(addresses, ep.iface.addrv6.String())
	}

	return sb.updateHostsFile(addresses)
}
```

Reply:
I ended up getting rid of `Endpoint.updateHostsFile()` ...

Instead, a new `Endpoint.getEtcHostsAddrs()` acquires the lock and does the combination of now-deleted `getFirstInterfaceIPv4Address()` and `getFirstInterfaceIPv6Address()`.

Then, the two Sandbox callers use that result to call `Sandbox.updateHostsFile()`.